### ISSUE TITLE
Avoid bind values in MERGE ON and VALUES claues

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11852]
Running a statement like this:

```java
assertEquals(1,
create().insertInto(TDates())
        .columns(TDates_ID(), TDates_TS())
        .values(1, t1)
        .onConflictDoNothing()
        .execute());
```

Produces an exception:

```
org.jooq.exception.DataAccessException: SQL [merge into "test"."t_dates" using (select 1 "one" from (select 1 as "dual") as "dual") on "test"."t_dates"."id" = 1 when not matched then insert ("id", "ts") values (?, ?)]; [Teradata Database] [TeraJDBC 16.20.00.06] [Error 5758] [SQLState HY000] The search condition must fully specify the Target table primary index and partition column(s) and expression must match INSERT specification primary index and partition column(s).
	at org.jooq_3.15.0-SNAPSHOT.TERADATA.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2871)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:661)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:346)
	at org.jooq.impl.AbstractDelegatingQuery.execute(AbstractDelegatingQuery.java:111)
	at org.jooq.test.all.testcases.InsertUpdateTests.testInsertOnConflictDoNothing(InsertUpdateTests.java:650)
	at org.jooq.test.jOOQAbstractTest.testInsertOnConflictDoNothing(jOOQAbstractTest.java:1604)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:768)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.sql.SQLException: [Teradata Database] [TeraJDBC 16.20.00.06] [Error 5758] [SQLState HY000] The search condition must fully specify the Target table primary index and partition column(s) and expression must match INSERT specification primary index and partition column(s).
	at com.teradata.jdbc.jdbc_4.util.ErrorFactory.makeDatabaseSQLException(ErrorFactory.java:309)
	at com.teradata.jdbc.jdbc_4.statemachine.ReceiveInitSubState.action(ReceiveInitSubState.java:103)
	at com.teradata.jdbc.jdbc_4.statemachine.StatementReceiveState.subStateMachine(StatementReceiveState.java:311)
	at com.teradata.jdbc.jdbc_4.statemachine.StatementReceiveState.action(StatementReceiveState.java:200)
	at com.teradata.jdbc.jdbc_4.statemachine.StatementController.runBody(StatementController.java:137)
	at com.teradata.jdbc.jdbc_4.statemachine.StatementController.run(StatementController.java:128)
	at com.teradata.jdbc.jdbc_4.TDStatement.executeStatement(TDStatement.java:389)
	at com.teradata.jdbc.jdbc_4.TDStatement.prepareRequest(TDStatement.java:576)
	at com.teradata.jdbc.jdbc_4.TDPreparedStatement.<init>(TDPreparedStatement.java:131)
	at com.teradata.jdbc.jdk6.JDK6_SQL_PreparedStatement.<init>(JDK6_SQL_PreparedStatement.java:30)
	at com.teradata.jdbc.jdk6.JDK6_SQL_Connection.constructPreparedStatement(JDK6_SQL_Connection.java:82)
	at com.teradata.jdbc.jdbc_4.TDSession.prepareStatement(TDSession.java:1337)
	at com.teradata.jdbc.jdbc_4.TDSession.prepareStatement(TDSession.java:1381)
	at com.teradata.jdbc.jdbc_4.TDSession.prepareStatement(TDSession.java:1367)
	at org.jooq.tools.jdbc.DefaultConnection.prepareStatement(DefaultConnection.java:95)
	at org.jooq.test.utils.LoggingConnection.prepareStatement(LoggingConnection.java:68)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.jdbc.datasource.SingleConnectionDataSource$CloseSuppressingInvocationHandler.invoke(SingleConnectionDataSource.java:325)
	at jdk.proxy2/jdk.proxy2.$Proxy31.prepareStatement(Unknown Source)
	at org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109)
	at org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:82)
	at org.jooq.impl.AbstractQuery.prepare(AbstractQuery.java:395)
	at org.jooq.impl.AbstractDMLQuery.prepare0(AbstractDMLQuery.java:834)
	at org.jooq.impl.AbstractDMLQuery.prepare(AbstractDMLQuery.java:825)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:299)
	... 34 more
```

It appears we can't have bind values there, in both `ON` and `VALUES`. Let's just inline them automatically, for the user.

**Comment by lukaseder** [comment]
While inlining the values fixes the problem, the actual problem here seems to be the `DUAL` emulation. This reproduces it:

```java
create().connection(c -> {
    try (PreparedStatement s = c.prepareStatement(
        """
        merge into "t_dates"
        using (select 1 "one" from (select 1 as "dual") as "dual")
        on "t_dates"."id" = ?
        when not matched then insert ("id", "ts") values (?, ?)
        """
    )) {
        s.setInt(1, 1);
        s.setInt(2, 1);
        s.setTimestamp(3, Timestamp.valueOf("2000-01-01 00:00:00"));
        assertEquals(1, s.executeUpdate());
        assertFalse(s.execute());
    }
});
```

But this does not:

```java
create().connection(c -> {
    try (PreparedStatement s = c.prepareStatement(
        """
        merge into "t_dates"
        using (select 1 a) t
        on "t_dates"."id" = ?
        when not matched then insert ("id", "ts") values (?, ?)
        """
    )) {
        s.setInt(1, 1);
        s.setInt(2, 1);
        s.setTimestamp(3, Timestamp.valueOf("2000-01-01 00:00:00"));
        assertEquals(1, s.executeUpdate());
        assertFalse(s.execute());
    }
});
```

Seems like a rather esoteric Teradata bug...

**Comment by lukaseder** [comment]
The fix applied for Derby in #5110 seems to also work here.

**Comment by lukaseder** [comment]
Commit included with https://github.com/jOOQ/jOOQ/issues/11850
